### ISSUE TITLE
Menu ignores video_rotation configuration

### ISSUE THREAD

**Issue by natinusala** [libretro/RetroArch#6770]
## Description

The menu doesn't rotate with the video_rotation configuration set to something else than 0

### Expected behavior

The menu should rotate as well as the cores, without any stretching issues.

### Actual behavior

Nothing happens for the menu, and the cores image is stretched.

### Steps to reproduce the bug

1. Open RetroArch in a menu
2. Set the video rotation to some degree in the options
3. See that the menu is not rotated
4. Open some content, see that it's stretched

### Bisect Results

This option was never meant to rotate the menu.

### Version/Commit
- RetroArch: d360265

### Environment information

- OS: Lakka on the Nintendo Switch (no xorg server)
- Compiler: Lakka build script

**Comment by gouchi** [comment]
Not sure if we can use this [code](https://github.com/jwrdegoede/plymouth/commit/2031c804753e217704e43ec986eb547a364b1aae) to do it ?

[Source](https://github.com/libretro/RetroArch/issues/6258)

**Comment by natinusala** [comment]
This code is about _knowing_ that the display should be rotated.

Here (Lakka on the Switch), we're targeting very specific hardware where we already know that the display should be rotated. The issue being : how should we rotate it ? For now, we can't.

**Comment by Huntereb** [comment]
This should be considered a fairly major issue in all regards, simply because physically rotating a display is a real thing. Hopefully it's not a majorly difficult thing to accomplish, but it's certainly impressive to see not already implemented!

**Comment by gouchi** [comment]
@natinusala Yes right it is just for knowing the orientation.

It seems there is [rotation kms properties](https://01.org/linuxgraphics/gfx-docs/drm/drm-kms-properties.html).

**Comment by bparker06** [comment]
There is now a `screen_orientation` setting for handling this.

**Comment by ofry** [comment]
@bparker06 Screen Orientation rotates all screen, not only Retroarch window content.

**Comment by fr500** [comment]
That's correct

**Comment by smokku** [comment]
As I cen see in the commits implementing this, it only work with X11 and Win32 drivers.

(I'm trying to configure rotation in Lakka-Generic.x86_64-2.2.2.img and it does not work.)

**Comment by gouchi** [comment]
@smokku it is working on X11 and on Lakka Generic we are using DRM/KMS not X11.

**Comment by smokku** [comment]
I tried to use it on latest Lakka on a PC with nVidia card and:
- the menu option is not visible in UI (with advanced options enabled)
- when changing `screen_orientation=1` by directly editing the config file - rotation did not change

**Comment by gouchi** [comment]
@smokku May you try with [image](http://nightly.builds.lakka.tv/2019-08-03/Generic_VK_nvidia.x86_64/Lakka-Generic_VK_nvidia.x86_64-2.2-devel-20190803001151-r28321-gc3108f0a77.img.gz).

**Comment by bparker06** [comment]
@gouchi Are you talking about the `screen_orientation` setting working in Lakka? Because I don't see a display server in RA for DRM, so I don't see how that could work.

**Comment by smokku** [comment]
I browsed briefly through DRM/KMS docs and there is [plane rotation property](https://01.org/linuxgraphics/gfx-docs/drm/gpu/drm-kms.html#c.drm_plane_create_rotation_property). I will look into adding support for `screen_orientation` to DRM/KMS diriver.

**Comment by natinusala** [comment]
The screen rotation property is just for information purposes - we still need to flip the coordinates and rotate the display inside of retroarch

**Comment by smokku** [comment]
I thought it looked too easy. ;-)

**Comment by gouchi** [comment]
@bparker06 the [image provided](https://www.lakka.tv/articles/2019/01/16/lakka-generic-nvidia-experimental-builds/) is special one for nVidia card which runs X11.

**Comment by smokku** [comment]
Any hints on workflow for developing DRM/KMS driver? When switch to VTY and run retroarch it works, but I have no keyboard controls - to the point I cannot switch back to X. I need to ssh to my machine and kill retroarch to regain keyboard. ;-)

**Comment by gouchi** [comment]
Are you in the [input group](https://github.com/libretro/RetroArch/issues/5033#issuecomment-450706696) ? 

You may try this [documentation](https://events.static.linuxfound.org/sites/events/files/slides/brezillon-drm-kms.pdf) about DRM/KMS. Also you can use [modetest](https://wiki.st.com/stm32mpu/wiki/DRM_KMS_overview#modetest_-28DRM-2FKMS_test_tool-29) from libdrm ([autotools](https://elinux.org/R-Car/Tests:KMS-modetest#Obtaining_modetest) | [meson](https://gitlab.freedesktop.org/mesa/drm/blob/master/meson_options.txt#L133)) to get display information and set resolution.

There is [this code in RA](https://github.com/libretro/RetroArch/blob/master/gfx/drivers/drm_gfx.c#L1018) but somebody more familiar will have better answer.

**Comment by smokku** [comment]
Thanks for the `/dev/input` tip.
I also needed `./configure --enable-udev`.

Now it works on VTY. :)

**Comment by smokku** [comment]
A bit more digging through code and I see that `screen_orientation` is a setting specific to a display server (currently: android, null, win32, x11) and uses display server facilities to rotate the whole display on the OS level.

The setting `video_rotation` is for core only, so even implementing it in `drm_gfx` driver would not help rotating the menu.

OK. This is a lot more involved.
Any information on why menu is not affected by `video_rotation` setting?

**Comment by natinusala** [comment]
Because it's not implemented

**Comment by smokku** [comment]
Is there a use case, when we would like to rotate the core, but not the menu?

**Comment by natinusala** [comment]
Arcade or Nintendo DS cores for instance

**Comment by smokku** [comment]
I do understand that there are games and emus that use vertical screen layout.
In fact, I've built an arcade cabinet with vertical screen, because I play mostly vertical shooters. That's why I am interested in working on this feature - I would like to use Lakka with my cabinet.

It puzzles me though for an use case when you would like the menu not rotated, but rotated game/core.
I guess for devices which you can easily rotate, like mobile phone, this could be useful to have horizontal menu and vertical game... IDK

**Comment by natinusala** [comment]
As I said, desmume has the ability to request a rotation to the frontend because whatever current content needs the display to be rotated - I'm sure it's not the only core

**Comment by unformatt** [comment]
Hard to follow this thread with regards to `screen_orientation` setting. Seems like it works on some systems and not others? I'm wondering why this is a Lakka or RA issue. Doesn't linux support setting the screen orientation?

**Comment by gouchi** [comment]
@unformatt We don't use Xorg by default Lakka  so you can't use `screen_orientation` but there is an [experimental version dedicated for Nvidia cards](https://www.lakka.tv/articles/2019/01/16/lakka-generic-nvidia-experimental-builds/) which is using Xorg so you could change screen orientation with this version.

**Comment by ChristopherHaynes** [comment]
Any further update on an implementation for menu rotation in with DRM/KMS?

I'm assuming there is an limitation within RetroArch which is making this a difficult feature to add.

**Comment by theofficialgman** [comment]
Sorry to revive a dead thread.

Due to the Nvidia Xorg Rotation Stutter Bug https://wiki.switchroot.org/wiki/common-issues#stutter-framepacing-issue-when-handheld we REALLY want to use "Video Rotation" rather than "Screen Orientation" (xorg).

I just re-discovered that the "Video Rotation" setting exists and this fixes the horrible framepacing that users have been dealing with when handheld (since its a portrait screen that traditionally we rotate with xorg, but if we instead leave screen orientation at "normal" (portrait) and "video rotation" at 90deg we can get gameplay to show as landscape in handheld). However, as noted in the initial issue message, the UI is not rotated. It would be great if there was an option to also rotate the UI by the same amount WITHOUT resorting to "Screen Orientation" xorg rotation (which has the horrible stutter/framepacing issue).
